src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,102 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final double CAPACITY_MULTIPLIER = 1.5;
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private T[] elementsData;
+    private int size;
+
+    public ArrayList() {
+        this.elementsData = (T[]) new Object[DEFAULT_CAPACITY];
+        this.size = 0;
+    }
+
+    private void indexExistence(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index);
+        }
+    }
+
+    private void resizeCheck(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index for add operation: "
+                    + index);
+        }
+    }
+
+    private void ensureCapacity(int minCapacity) {
+        if (minCapacity > elementsData.length) {
+            int newCapacity = (int) (elementsData.length * CAPACITY_MULTIPLIER);
+            elementsData = Arrays.copyOf(elementsData, Math.max(newCapacity, minCapacity));
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
+        elementsData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        resizeCheck(index);
+        ensureCapacity(size + 1);
+        System.arraycopy(elementsData, index, elementsData, index + 1, size - index);
+        elementsData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        ensureCapacity(size + list.size());
```

Comment:
You don't need to call this method; it's included in the add() method that you are calling in iteration.

Reply:
it is fine to call it here, because we check for the whole list size at once, but the fact that it is also being called in the `add(T value)`.. I think I would rather add a custom logic of adding a whole list instead of reusing `add(T value)` because of this, and this should make it more efficient, but overall this is not a big deal if you were to leave it as it is